### PR TITLE
v.decimate: Added tests

### PR DESCRIPTION
**Pull request by dvrohan**
Adds tests covering different parameters of the tool, although testing only the basics for the outputs.

### CODE REVIEW

**Comment by dvrohan** [comment]
The tests is running fine on my machine. I am using a Linux distro. More specifically WSL2 on Windows. However it fails in the MACOS and Windows test here. Could someone help me with this?

**Comment by echoix** [comment]
On macOS I'm not sure what the cause is from the trace. On Windows there's a clue about threading and lock. So, I assume there's something to do with multiprocessing.

On Linux, if you use Python 3.14 rc1 (that means it is binary compatible and projects are encouraged to start building wheels), they should be using the same start mechanisms that are used on macOS and Windows (safer, but sometimes code needs to be adjusted). I didn't find something that was problematic in the new code, so maybe take a look at the tool implementation (if it is written in Python).

So, if you're able to use Python 3.14 on Linux, and it fails the same way, and you manage to fix it on Linux, there's a big chance that it will be fixed on both other platforms.

**Comment by dvrohan** [comment]
I was initially using python 3.10. I install python 3.14 rc1. The tests are still running fine with this version as well.

**Comment by wenzeslaus** [comment]
v.decimate is in C, but the v.decimate process exits with 3221225477 (0xC0000005) on Windows which is some kind of system error (I can't find official doc for that). macOS ends with an error coming from GRASS, "No attribute table for layer 1". That's specifically from v.decimate which for some reason expects attribute table, but does not find one. Finally, Linux works. This may indicate error in v.decimate code, but Coverity Scan and GitHub CodeQL do not should anything. Running the tool with Valgrind would be my next step. For the tests, I suggest to remove the problematic test from this PR to merge what works and open a new PR only with the failing test afterwards.

File: vector/v.decimate/tests/test_v_decimate.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+
+    output_map = f"decimated_preserve_{preserve_value}"
+
+    tools.v_decimate(
+        input=input_map, output=output_map, preserve=preserve_value, overwrite=True
+    )
+
+    result = tools.g_list(type="vector", pattern=output_map, format="json")
+    names = [entry["name"] for entry in result]
+    assert output_map in names, f"Decimated map {output_map} not found in vector list"
```
Comment (wenzeslaus) [inline vector/v.decimate/tests/test_v_decimate.py]: Testing number of points with v.info is more appropriate than just testing map presence, even testing that there is non-zero number of points or that some other property  is correct (like v.decimate is mentioned in the history) would be better. The `decimated < original` test would be a good starting point, although further tests for exact count would be nice too.

File: vector/v.decimate/tests/test_v_decimate.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    input_map, session = setup_point_map
+    tools = Tools(session=session)
+
+    output_map = "decimated_skip_3"
+    tools.v_decimate(input=input_map, output=output_map, skip=3, overwrite=True)
+
+    result = tools.g_list(type="vector", pattern=output_map, format="json")
+    assert any(v["name"] == output_map for v in result), (
+        f"Decimated map {output_map} not found"
+    )
```
Comment (wenzeslaus) [inline vector/v.decimate/tests/test_v_decimate.py]: Same as above.

File: vector/v.decimate/tests/test_v_decimate.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+            overwrite=True,
+        )
+
+        yield mapname, session
+
+
+def _get_point_count(mapname, tools):
+    """Helper to get number of points in a vector map."""
+    info = tools.v_info(map=mapname, flags="t", format="json")
+    return info["points"]
```
Comment (wenzeslaus) [inline vector/v.decimate/tests/test_v_decimate.py]: Maybe matter of taste, but I don't think this helps readability. Elsewhere, you need to use the full result anyway, so no wrapper there.

File: vector/v.decimate/tests/test_v_decimate.py (hunk 2 above)
Comment (wenzeslaus) [reply vector/v.decimate/tests/test_v_decimate.py]: Also below...

File: vector/v.decimate/tests/test_v_decimate.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    with pytest.raises(CalledModuleError) as excinfo:
+        tools.v_decimate(
+            input=input_map,
+            output="fail_test",
+            flags="g",  # -g requires at least one of -f, -c, cell_limit, -z, zdiff
+            overwrite=True,
+        )
+    assert "requires at least one of" in str(excinfo.value), (
+        "Expected error about missing required flags or parameters"
+    )
```
Comment (wenzeslaus) [inline vector/v.decimate/tests/test_v_decimate.py]: There is a matches parameter for the raises function.

File: vector/v.decimate/tests/test_v_decimate.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+        tools.g_region(n=110, s=90, e=110, w=90, res=1)
+
+        # Import points as 3D vector map (z flag and z=3 column)
+        tools.v_in_ascii(
+            input=io.StringIO(coords),
+            output=mapname,
+            separator="|",
+            flags="z",
+            z=3,
+            overwrite=True,
```
Comment (wenzeslaus) [inline vector/v.decimate/tests/test_v_decimate.py]: No overwrite should be needed in this code.

File: vector/v.decimate/tests/test_v_decimate.py [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+        )
+
+        yield mapname, session
```
Comment (wenzeslaus) [inline vector/v.decimate/tests/test_v_decimate.py]: I had at least some success using a SimpleNamespace object here. It saves the assignment `input_map, session = setup_point_map` in each function as using the attributes instead makes it easier than the tuple.

File: vector/v.decimate/tests/test_v_decimate.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+    tools = Tools(session=session)
+    output_map = "decimated_all_layers"
+
+    tools.v_decimate(
+        input=input_map,
+        output=output_map,
+        layer=-1,  # All layers
+        skip=2,
+        overwrite=True,
+    )
```
Comment (wenzeslaus) [inline vector/v.decimate/tests/test_v_decimate.py]: You don't setup more than one layer, so either remove this test, add more complex test data, or simply document explicitly that you are selecting all layers while only one layer is present.

**Review by wenzeslaus — CHANGES_REQUESTED** [review]

File: vector/v.decimate/tests/test_v_decimate.py (hunk 1 above)
Comment (dvrohan) [reply vector/v.decimate/tests/test_v_decimate.py]: Implemented the test as mentioned.

File: vector/v.decimate/tests/test_v_decimate.py (hunk 2 above)
Comment (dvrohan) [reply vector/v.decimate/tests/test_v_decimate.py]: Changed tests accordingly

File: vector/v.decimate/tests/test_v_decimate.py (hunk 3 above)
Comment (dvrohan) [reply vector/v.decimate/tests/test_v_decimate.py]: Agreed. I have just called the v_info function in the respective functions.

File: vector/v.decimate/tests/test_v_decimate.py (hunk 4 above)
Comment (dvrohan) [reply vector/v.decimate/tests/test_v_decimate.py]: Used the match parameter

File: vector/v.decimate/tests/test_v_decimate.py (hunk 5 above)
Comment (dvrohan) [reply vector/v.decimate/tests/test_v_decimate.py]: Removed the overwrite.

File: vector/v.decimate/tests/test_v_decimate.py (hunk 7 above)
Comment (dvrohan) [reply vector/v.decimate/tests/test_v_decimate.py]: Removed the test to be implemented for further PR.

File: vector/v.decimate/tests/test_v_decimate.py (hunk 6 above)
Comment (dvrohan) [reply vector/v.decimate/tests/test_v_decimate.py]: I have updated the code to use SimpleNamespace

File: vector/v.decimate/tests/test_v_decimate.py [hunk 8]
```diff
@@ -8,225 +10,198 @@
...
     tools.v_decimate(input=input_map, output=output_map, skip=3, overwrite=True)
 
-    result = tools.g_list(type="vector", pattern=output_map, format="json")
-    assert any(v["name"] == output_map for v in result), (
-        f"Decimated map {output_map} not found"
+    orig_info = tools.v_info(map=input_map, flags="t", format="json")
+    dec_info = tools.v_info(map=output_map, flags="t", format="json")
+
+    orig_count = int(orig_info["points"])
+    dec_count = int(dec_info["points"])
```
Comment (wenzeslaus) [inline vector/v.decimate/tests/test_v_decimate.py]:
```suggestion
    orig_count = tools.v_info(map=input_map, flags="t", format="json")["points"]
    dec_count = tools.v_info(map=output_map, flags="t", format="json")["points"]
```

File: vector/v.decimate/tests/test_v_decimate.py [hunk 9]
```diff
@@ -8,225 +10,198 @@
...
     tools = Tools(session=session)
     output_map = "out_offset_limit"
 
     tools.v_decimate(
         input=input_map, output=output_map, offset=2, limit=3, overwrite=True
     )
 
-    count = _get_point_count(output_map, tools)
+    info = tools.v_info(map=output_map, flags="t", format="json")
+    count = info["points"]
```
Comment (wenzeslaus) [inline vector/v.decimate/tests/test_v_decimate.py]:
```suggestion
    count = tools.v_info(map=output_map, flags="t", format="json")["points"]
```

File: vector/v.decimate/tests/test_v_decimate.py [hunk 10]
```diff
@@ -8,225 +10,198 @@
...
-
-    # Set region with 1m resolution to create grid cells
-    tools.g_region(vector=input_map, res=1.0)
-
-    # Decimate using grid with z-difference threshold (zdiff)
-    tools.v_decimate(
-        input=input_map, output=output_map, flags="gz", zdiff=0.5, overwrite=True
-    )
+    info = tools.v_info(map=output_map, flags="t", format="json")
+    count = info["points"]
```
Comment (wenzeslaus) [inline vector/v.decimate/tests/test_v_decimate.py]:
```suggestion
    count = tools.v_info(map=output_map, flags="t", format="json")["points"]
```
